### DISCUSSION TITLE
IGMPv2 issue in response to IGMPv3  BLOCK_OLD_SOURCES on pimd 2.3.2

### DISCUSSION THREAD

**Discussion by franciszekk** [troglobit/pimd#247] [General]
Opening this discussion related to pimd version **2.3.2**.
In this version the service in response to igmpv3 message that include "IGMP_BLOCK_OLD_SOURCES" sends to the network IGMP_MEMBERSHIP_QUERY  igmp**v3** (src:  [igmp_proto.c accept_leave_message() line 520](https://github.com/troglobit/pimd/blob/6b91405516e9f919f2df23f7794d00e916fc1871/igmp_proto.c#L520))  and next second IGMP_MEMBERSHIP_QUERY but this time as igmp**v2** even if the debug log is printing igmp**v3**

After investigation, found that this is due to implementation  of _SendQuery()_  [igmp_proto.c SendQuery() line 878](https://github.com/troglobit/pimd/blob/6b91405516e9f919f2df23f7794d00e916fc1871/igmp_proto.c#L878)  which is invoked by a timer, this  function call _send_query()_  omitting the variable value _cbk->q_len_ causing it to send an igmp**v2** IGMP_MEMBERSHIP_QUERY  (src: [gmp_proto.c send_query() line 866](https://github.com/troglobit/pimd/blob/6b91405516e9f919f2df23f7794d00e916fc1871/igmp_proto.c#L866) ). 

The side effect of this behaviour is that if the client, which expects SSM multicast, is changing the multicast source ip, receive IGMP_MEMBERSHIP_QUERY igmpv3 followed by another  IGMP_MEMBERSHIP_QUERY igmp**v2**, In response to this sequence, the client switches igmp**v2** compatibility mode (see the linux igmp implementation, the usage of macro [IGMP_V2_SEEN](https://github.com/torvalds/linux/blob/2c8159388952f530bd260e097293ccc0209240be/net/ipv4/igmp.c#L130)   relies on the variable _mr_v2_seen_)

I see that on master branch, thanks to the commit  [Refactor, sync with pimd-dense](https://github.com/troglobit/pimd/commit/f6864604f9786e6b309f30623c019cc8d4f9fe00)  this is changed, now in SendQuery() the value _cbk->q_len_ is not omitted but passed to the function _send_igmp()_ 

Question if this commit can be backported to pimd 2.3.2 (for various reason build pimd from maser is not possible) ?

**Comment by troglobit** [comment]
There are several fixes on the master branch that improve on the state of pimd v2.3.2.  The idea was to create a release v3.0, but due to lack of time on my part and community engagement this has not happened yet.

For the same reasons there will also not be any v2.3.3 with backports of select fixes.
